lib/pure/algorithm.nim

```diff
@@ -369,3 +369,113 @@ when isMainModule:
   for i in 0 .. high(arr1):
     assert arr1.reversed(0, i) == arr1.reversed()[high(arr1) - i .. high(arr1)]
     assert arr1.reversed(i, high(arr1)) == arr1.reversed()[0 .. high(arr1) - i]
+
+
+proc rotate*[T](arg: var openarray[T]; first, middle, last: int): int {.discardable.} =
+  ## Performs a left rotation on a range of elements.
+  ## Specifically, ``rotate`` swaps the elements in the range [first, last) in such a way
+  ## that the element at index ``middle`` becomes the first element of the sub range and
+  ## ``middle - 1`` becomes the last element of the sub range. The time complexity is
+  ## linear to ``last - first``. returns ``first + (last - middle)``. ``T`` must support swap operation
+  ##
+  ## ``first``
+  ##   the index of the first element that should be rotated.
+  ##
+  ## ``middle``
+  ##   the index of the element that should appear at the beginning of the rotated range.
+  ##
+  ## ``last``
+  ##   the index of the first element that should not be rotated anymore (exclusive upper bound).
```

Comment:
We don't use exclusive upper bounds in Nim's stdlib. They don't work when you don't have an element past the last element you want to be included (what is ``high(myenum)+1``?).